src/shared/copy.c

```diff
@@ -657,7 +657,10 @@ static int hardlink_context_setup(
                         return -errno;
         }
 
-        r = tempfn_random_child(to, "hardlink", &c->subdir);
+        if (to)
+                r = tempfn_random_child(to, "hardlink", &c->subdir);
+        else
+                r = tempfn_random("hardlink", /* extra= */ NULL, &c->subdir);
```

Comment:
Isn't this extremely broken already?

`to` is just a filename, not a full path, `tempfn_random_child()` should take the given directory file descriptor `dt` into account as well, not just `to`.

Any chance you can add `tempfn_random_child_at()` and then use it here for both cases?